--- mod_rpaf-0.5.orig/mod_rpaf-2.0.c	2004-03-17 17:47:30.000000000 +0200
+++ mod_rpaf-0.5/mod_rpaf-2.0.c	2007-09-27 13:23:45.335194758 +0300
@@ -130,6 +130,16 @@
     return 0;
 }
 
+static char* last_not_in_array(apr_array_header_t *forwarded_for,
+			       apr_array_header_t *proxy_ips) {
+    int i;
+    for (i = (forwarded_for->nelts)-1; i > 0; i--) {
+	if (!is_in_array(((char **)forwarded_for->elts)[i], proxy_ips))
+	    break;
+    }
+    return ((char **)forwarded_for->elts)[i];
+}
+
 static int change_remote_ip(request_rec *r) {
     const char *fwdvalue;
     char *val;
@@ -147,7 +157,7 @@
                 if (*fwdvalue != '\0')
                     ++fwdvalue;
             }
-            r->connection->remote_ip = apr_pstrdup(r->connection->pool, ((char **)arr->elts)[((arr->nelts)-1)]);
+            r->connection->remote_ip = apr_pstrdup(r->connection->pool, last_not_in_array(arr, cfg->proxy_ips));
             r->connection->remote_addr->sa.sin.sin_addr.s_addr = inet_addr(r->connection->remote_ip);
             if (cfg->sethostname) {
                 const char *hostvalue;
--- mod_rpaf-0.5.orig/mod_rpaf.c	2007-09-27 13:23:30.864865219 +0300
+++ mod_rpaf-0.5/mod_rpaf.c	2007-09-27 13:23:59.885526119 +0300
@@ -130,6 +130,16 @@
     return 0;
 }
 
+static char* last_not_in_array(array_header *forwarded_for,
+			       array_header *proxy_ips) {
+    int i;
+    for (i = (forwarded_for->nelts)-1; i > 0; i--) {
+	if (!is_in_array(((char **)forwarded_for->elts)[i], proxy_ips))
+	    break;
+    }
+    return ((char **)forwarded_for->elts)[i];
+}
+
 static int change_remote_ip(request_rec *r) {
     const char *fwdvalue;
     char *val;
@@ -147,7 +157,7 @@
                 if (*fwdvalue != '\0')
                     ++fwdvalue;
             }
-            r->connection->remote_ip = ap_pstrdup(r->connection->pool, ((char **)arr->elts)[((arr->nelts)-1)]);
+            r->connection->remote_ip = ap_pstrdup(r->connection->pool, last_not_in_array(arr, cfg->proxy_ips));
 			switch (r->connection->remote_addr.ss_family) {
 			case AF_INET:
 				inet_pton(AF_INET, r->connection->remote_ip, &((struct sockaddr_in*)&r->connection->remote_addr)->sin_addr.s_addr);
